### PR TITLE
feat: multichain token detection

### PR DESCRIPTION
**Pull request by salimtb**
## **Description**
TokenDetectionController is responsible for detecting and keeping an updated list of all tokens across supported chains. This dataset is stored in the detectedTokens state variable within Metamask’s state. After completing this task, token detection will be enhanced by implementing periodic polling across all networks linked to the wallet, resulting in a more comprehensive dataset available to users.

For each network added to the wallet, the polling loop will receive the network as a parameter and execute token autodetection for it. Once results are available, they will be stored in detectedTokensAllChains, organized by chainId. This approach enables us to retrieve a comprehensive list of detected tokens across all networks.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/28258?quickstart=1)

## **Related issues**

Fixes: [#3431](https://github.com/MetaMask/MetaMask-planning/issues/3431)

## **Manual testing steps**

1. install dependencies using `yarn`
2. start the project using` PORTFOLIO_VIEW=1 yarn start`
3. add other networks where you have tokens
4. the autodetection should be multichain

### **After**

https://github.com/user-attachments/assets/a0db910c-b3a3-456d-b1b9-0c6a3e472976

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by salimtb** [comment]
closing this PR because the changes has been moved to [this](https://github.com/MetaMask/metamask-extension/pull/28444) one

**Comment by salimtb** [comment]
closing this PR because the changes has been moved to https://github.com/MetaMask/metamask-extension/pull/28444 one

**Comment by gambinish** [comment]
I'm noticing an unexpected behavior. When I am on a network, and I try to import my account's tokens, in the detected-token popover, I only see balances and tokenFiatBalance of the detected tokens on the chain that I am on.

Is this expected? I think we have token balances being polled cross-chain, so I wonder if it would be worth it to integrate these cross chain balances on this PR as well.

https://github.com/user-attachments/assets/eff54551-4d9d-4484-a452-8c60ecacea43

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -4082,14 +4082,6 @@ export default class MetamaskController extends EventEmitter {
       tokenListStopPollingByPollingToken:
         tokenListController.stopPollingByPollingToken.bind(tokenListController),
 
-      tokenBalancesStartPolling: tokenBalancesController.startPolling.bind(
```
Comment (sahar-fehri) [inline app/scripts/metamask-controller.js:4085]: could you help me understand why this had to be deleted? 🙏

**Comment by gambinish** [comment]
I'm noticing two bugs, that we've discussed on Slack. Posting them here for tracking:

1. On a newly imported wallet, I cannot import from my previously existing networks (this could be an existing behavior). After adding a new network, the tokens on my previously added chains get detected:

https://github.com/user-attachments/assets/355ab683-69cc-42dd-88db-14f02499fa45

2. There are certain scenarios where switching networks and toggling between network filter options pollutes the detected tokens, resulting in improperly imported tokens:

https://github.com/user-attachments/assets/c3c94561-a250-43ce-bc1d-009d7068105e

**Review by gambinish — DISMISSED** [review]
Pulled down, tested, LGTM. The bugs I reported earlier now seem to be resolved ✅ 

Once https://github.com/MetaMask/metamask-extension/pull/28593 gets through the merge queue, I think we can just rebase to develop and get these changes in 🚀

**Review by gambinish — DISMISSED** [review]

**Review by vinnyhoward — DISMISSED** [review]

File: ui/components/app/detected-token/detected-token.js [hunk 2]
```diff
@@ -79,22 +134,53 @@ const DetectedToken = ({ setShowDetectedTokens }) => {
...
+          dispatch(setNewTokensImported(tokenSymbols.join(', ')));
+        },
+      );
+
+      await Promise.all(importPromises);
+    } else {
+      await dispatch(addImportedTokens(selectedTokens, networkClientId));
+      const tokenSymbols = selectedTokens.map(({ symbol }) => symbol);
+      dispatch(setNewTokensImported(tokenSymbols.join(', ')));
+    }
```
Comment (NidhiKJha) [inline ui/components/app/detected-token/detected-token.js:173]: Nit: Can we move it to a selector as a follow up task?

**Review by NidhiKJha — COMMENTED** [review]
With the feature flag on, it looks really good. But I am seeing the detection even without feature flag when I run it with `yarn start`


https://github.com/user-attachments/assets/c44e79e3-0ac0-4c42-839b-fc1271ec5166

File: ui/components/app/detected-token/detected-token-details/detected-token-details.js [hunk 3]
```diff
@@ -39,8 +39,7 @@ const DetectedTokenDetails = ({
         badge={
           <AvatarNetwork
             size={AvatarNetworkSize.Xs}
-            name={currentNetwork?.nickname || ''}
-            src={currentNetwork?.rpcPrefs?.imageUrl}
+            src={CHAIN_ID_TO_NETWORK_IMAGE_URL_MAP[chainId]}
```
[the remark this answers is not in the document]
Comment (salimtb) [reply ui/components/app/detected-token/detected-token-details/detected-token-details.js:43]: yes because we don't rely on the current chain anymore

**Review by gambinish — APPROVED** [review]

**Review by NidhiKJha — APPROVED** [review]

**Comment by sahar-fehri** [comment]
Hey @salimtb Im seeing this behavior where after hiding a token and going back to main page i dnt see the same token list?

https://github.com/user-attachments/assets/bc91e987-ee79-4da1-acf8-f81c54ece46e

**Review by sahar-fehri — APPROVED** [review]

**Pull request merged**